src/puppetlabs/puppetdb/jdbc.clj

```diff
@@ -289,40 +287,27 @@
      (fn []
        ~@body)))
 
-(defn with-query-results-cursor*
-  "Executes the given parameterized query within a transaction,
-  producing a lazy sequence of rows. The callback `func` is executed
-  on the entire sequence.
-
-  The lazy sequence is backed by an active database cursor, and is thus
-  useful for streaming very large resultsets.
-
-  The cursor is closed when `func` returns. If an exception is thrown,
-  the query is cancelled."
-  [func sql params]
-  (sql/transaction
-   (with-open [stmt (.prepareStatement (sql/connection) sql)]
-     (doseq [[index value] (map vector (iterate inc 1) params)]
-       (.setObject stmt index value))
-     (.setFetchSize stmt 500)
-     (with-open [rset (.executeQuery stmt)]
-       (try
-         (-> rset
-             (sql/resultset-seq)
-             (convert-result-arrays)
-             (func))
-         (catch Exception e
-           ;; Cancel the current query
-           (.cancel stmt)
-           (throw e)))))))
-
-(defmacro with-query-results-cursor
-  "Executes the given parameterized query within a transaction.
-  `body` is then executed with `rs-var` bound to the lazy sequence of
-  resulting rows. See `with-query-results-cursor*`."
-  [sql params rs-var & body]
-  `(let [func# (fn [~rs-var] (do ~@body))]
-     (with-query-results-cursor* func# ~sql ~params)))
+(defn with-query-results-cursor
+  [[sql & params]]
+  (let [stmt (.prepareStatement (sql/connection) sql)]
+    (doseq [[index value] (map-indexed vector params)]
+      (.setObject stmt (inc index) value))
+    (.setFetchSize stmt 500)
+    (let [rset (.executeQuery stmt)
+          closing-fn (fn [r]
+                       (do (.close stmt) (.close rset))
+                       r)
+          ;; TODO collapse this with convert-result-arrays
+          convert-arrays (map (partial kitchensink/mapvals
+                                       (comp #(cond-> % (kitchensink/array? %) vec)
+                                             #(cond-> % (isa? (class %) java.sql.Array) .getArray))))]
+      (try
+        (->> (sql/resultset-seq rset)
+             (eduction convert-arrays #(completing % closing-fn))
+             sequence)
```

Comment:
@senior so a few questions here, 1) this sequence call is definitely needed, tests break without it